File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

secToYears :: Fractional a => a -> a
secToYears sec = sec / 31557600

period :: Planet -> Float
period Mercury = 0.2408467
period Venus   = 0.61519726
period Earth   = 1
period Mars    = 1.8808158
period Jupiter = 11.862615
period Saturn  = 29.447498
period Uranus  = 84.016846
period Neptune = 164.79132

ageOn :: Planet -> Float -> Float
ageOn planet age = secToYears age / period planet

File: fsharp/space-age/SpaceAge.fs
module SpaceAge

type Planet = float
let Earth = 31557600.0
let Mercury = 0.2408467*Earth
let Venus = 0.61519726*Earth
let Mars = 1.8808158*Earth
let Jupiter = 11.862615*Earth
let Saturn = 29.447498*Earth
let Uranus = 84.016846*Earth
let Neptune = 164.79132*Earth

let age (planet: Planet) (seconds: int64): float = 
    float seconds / float planet